### PR TITLE
Add ReadCache slice operation

### PR DESCRIPTION
**Pull request by jcharum**
Add a `ReadCache` slice operation that reads from an existing cache but does not write any cache itself. This may be useful if you want to reuse a cache from a previous computation and fail if it does not exist.

This change also removes the errors returned by the `Cache` and `CachePartial` operations, as they are never non-`nil`. They previously were used to signal an incomplete cache, but this is now handled by the `IsCached` method. This simplifies callers, especially since these are virtually always called within `bigslice.Func` definitions which do not an easy way to propagate errors.

### CODE REVIEW

File: cache_test.go [hunk 1]
```diff
@@ -328,6 +310,78 @@ func TestCacheErr(t *testing.T) {
...
+		var v1, v2 int
+		ok := scan1.Scan(ctx, &v1)
+		if got, want := scan2.Scan(ctx, &v2), ok; got != want {
+			t.Errorf("got %v, want %v", got, want)
+			break
+		}
+		if !ok {
+			break
+		}
+		if v1 != v2 {
```
Comment (josh-newman) [inline cache_test.go]: It seems like this test depends on the local executor scanning shards in consistent order, if I'm not misreading. Is that property guaranteed? Either way, it seems pretty subtle, and maybe worth a comment. Or just scan both separately, then sort and compare.

File: cache.go [hunk 2]
```diff
@@ -62,10 +60,35 @@ func Cache(ctx context.Context, slice Slice, prefix string) (Slice, error) {
...
+func (*readCacheSlice) Dep(i int) Dep            { panic("no deps") }
+func (*readCacheSlice) Combiner() slicefunc.Func { return slicefunc.Nil }
+
+func (r *readCacheSlice) Reader(shard int, _ []sliceio.Reader) sliceio.Reader {
+	return r.cache.CacheReader(shard)
+}
+
+// ReadCache reads from an existing cache but does not write any cache itself.
+// This may be useful if you want to reuse a cache from a previous computation
+// and fail if it does not exist. typ is the type of the cached and returned
```
[the remark this answers is not in the document]
Comment (jcharum) [reply cache.go]: Do people realize that `slicetype.Type` is embedded in the `bigslice.Slice` interface? That is, you would probably just pass the slice you would cache in here and probably just call `slice.NumShard()` for the next argument.

I avoided having it take a `Slice` to try to make it clear that the result `Slice` would have no dependencies (and make the implementation also more clear), but maybe it's instead confusing?

File: cache_test.go [hunk 3]
```diff
@@ -328,6 +310,78 @@ func TestCacheErr(t *testing.T) {
...
+	}
+	if err := scan2.Err(); err != nil {
+		t.Errorf("scan2: %v", err)
+	}
+}
+
+// TestReadCacheError verifies that a ReadCache reader returns an error if the
+// cache does not exist.
+func TestReadCacheError(t *testing.T) {
+	dir, cleanup := testutil.TempDir(t, "", "")
```
[the remark this answers is not in the document]
Comment (jcharum) [reply cache_test.go]: This spelling is carried from the original test code from way back when. I agree, but I'll do it in a separate PR.

File: internal/slicecache/slicecache.go [hunk 4]
```diff
@@ -93,10 +95,17 @@ func (c *FileShardCache) WritethroughReader(shard int, reader sliceio.Reader) sl
...
 
-// CacheReader returns a reader that reads from the cache.
+// CacheReader returns a reader that reads from the cache. If the shard is not
+// cached, returns a reader that will always return an error.
 func (c *FileShardCache) CacheReader(shard int) sliceio.Reader {
 	if !c.shardIsCached[shard] {
-		return sliceio.ErrReader(fmt.Errorf("shard %d is not cached", shard))
+		path := c.path(shard)
+		if c.requireAll {
+			path = fmt.Sprintf("%s-NNNN-of-%04d", c.prefix, c.numShards)
```
Comment (josh-newman) [inline internal/slicecache/slicecache.go]: Optional: Declare this and `path`'s format strings in a top-level block, together, to make sure we don't accidentally change one without the other.

File: cache.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (josh-newman) [reply cache.go]: Actually, I think it's a good idea to not pass a full `Slice`, I think it's important the no-dependencies is clear and that's the best way to do it. I think people will figure out the `Type` thing easily enough.

**Pull request merged**